File: android/app/src/main/java/com/laneshadow/ui/components/molecules/FavoritesInfoSheet.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.ui.components.atoms.Button
import com.laneshadow.ui.components.atoms.ButtonSize
import com.laneshadow.ui.components.atoms.ButtonVariant
import com.laneshadow.ui.components.atoms.IconSymbol
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * FavoritesInfoSheet molecule component
 *
 * Bottom sheet dialog for displaying informational message when favorite roads
 * couldn't be included in route planning due to distance constraints.
 * Following React Native wrapper patterns from react-native/components/sheets/favorites-info-sheet.tsx
 *
 * @param visible Whether to show the info sheet
 * @param onClose Callback when user taps "Got it" button or dismisses the sheet
 * @param unavailableFavorites List of favorite road names that couldn't be included
 * @param testID Optional test ID for UI testing
 */
@Composable
fun FavoritesInfoSheet(
    visible: Boolean,
    onClose: () -> Unit,
    unavailableFavorites: List<String>,
    testID: String? = null,
) {
    if (!visible) {
        return
    }

    val theme = LocalLaneShadowTheme.current

    // Build accessibility description
    val accessibilityDescription = "Favorites not included. These favorite roads are too far from your planned route: ${unavailableFavorites.joinToString(", ")}"

    AlertDialog(
        onDismissRequest = onClose,
        modifier = Modifier
            .testTag(testID ?: "favorites-info-sheet")
            .semantics {
                contentDescription = accessibilityDescription
            },
        icon = {
            // Info icon in circular container (primary color at 15% opacity)
            Box(
                modifier = Modifier
                    .size(64.dp)
                    .background(
                        color = theme.colors.primary.default.copy(alpha = 0.15f),
                        shape = CircleShape,
                    )
                    .padding(theme.space.md)
                    .testTag(testID?.let { "$it-icon" } ?: "favorites-info-icon"),
                contentAlignment = Alignment.Center
            ) {
                IconSymbol(
                    name = "information",
                    size = 32.dp,
                    color = theme.colors.primary.default,
                )
            }
        },
        title = {
            // Title: "Favorites Not Included" (title.md = 20sp, SemiBold, centered)
            Text(
                text = "Favorites Not Included",
                color = theme.colors.onSurface.default,
                style = theme.type.title.md,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag(testID?.let { "$it-title" } ?: "favorites-info-title")
            )
        },
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = theme.space.lg),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(theme.space.md)
            ) {
                // Message (body.md, muted color)
                Text(
                    text = "These favorite roads are too far from your planned route:",
                    color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                    style = theme.type.body.md,
                    textAlign = TextAlign.Center,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag(testID?.let { "$it-message" } ?: "favorites-info-message")
                )

                // Favorites list (surface bg at 50% opacity, md radius, md padding, sm gap)
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(
                            color = theme.colors.surface.default.copy(alpha = 0.5f),
                            shape = androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.md),
                        )
                        .padding(theme.space.md)
                        .testTag(testID?.let { "$it-list" } ?: "favorites-info-list")
                ) {
                    Column(
                        verticalArrangement = Arrangement.spacedBy(theme.space.sm)
                    ) {
                        unavailableFavorites.forEach { favorite ->
                            Text(
                                text = "• $favorite",
                                color = theme.colors.onSurface.default,
                                style = theme.type.body.md,
                                modifier = Modifier.testTag("favorite-item-$favorite")
                            )
                        }
                    }
                }

                // Guidance text (body.sm, muted color)
                Text(
                    text = "Try planning a route nearer to these favorites, or add them to a different route.",
                    color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                    style = theme.type.body.sm,
                    textAlign = TextAlign.Center,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag(testID?.let { "$it-guidance" } ?: "favorites-info-guidance")
                )
            }
        },
        confirmButton = {
            // Got it button (default variant, lg size, full width)
            Button(
                variant = ButtonVariant.Default,
                size = ButtonSize.Large,
                text = "Got it",
                onPress = onClose,
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag(testID?.let { "$it-close-button" } ?: "favorites-info-close-button"),
            )
        },
        containerColor = theme.colors.surface.default,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.xl),
    )
}

File: ios/LaneShadow/Views/Molecules/FavoritesInfoSheet.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Favorites Info Sheet Component

/**
 * Favorites info sheet molecule component
 *
 * Shows informational message when favorite roads couldn't be included in route.
 * Following React Native component from react-native/components/sheets/favorites-info-sheet.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Icon background: `theme.colors.primary.default` (15% opacity)
 *   - Icon: `theme.colors.primary.default`
 *   - Title: `theme.colors.onSurface.default`
 *   - Message/guidance: `theme.colors.onSurface.muted` (60% opacity)
 *   - List background: `theme.colors.surface.default` (50% opacity)
 *   - List items: `theme.colors.onSurface.default`
 * - Layout:
 *   - Container padding: `theme.space.lg`
 *   - Gap between elements: `theme.space.md`
 *   - List container padding: `theme.space.md`
 *   - List item gap: `theme.space.sm`
 *   - Icon container padding: `theme.space.md`
 *   - Icon radius: `theme.radius.full`
 *   - List container radius: `theme.radius.md`
 * - Typography:
 *   - Title: `theme.type.title.md` (medium weight, 18pt)
 *   - Message: `theme.type.body.md` (16pt)
 *   - List items: `theme.type.body.md` (16pt)
 *   - Guidance: `theme.type.body.sm` (14pt)
 *
 * ## Parameters
 * - visible: Whether to show the sheet
 * - unavailableFavorites: Array of favorite names that couldn't be included
 * - onDismiss: Callback when user dismisses the sheet
 * - testID: Optional test identifier for UI testing
 */
public struct LSFavoritesInfoSheet: View {
    @Environment(\.theme) private var theme

    private let visible: Bool
    private let unavailableFavorites: [String]
    private let onDismiss: () -> Void
    private let testID: String

    public init(
        visible: Bool,
        unavailableFavorites: [String],
        onDismiss: @escaping () -> Void,
        testID: String = "favorites-info-sheet"
    ) {
        self.visible = visible
        self.unavailableFavorites = unavailableFavorites
        self.onDismiss = onDismiss
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        EmptyView()
            .sheet(isPresented: .constant(visible)) {
                sheetContent
            }
            .accessibilityIdentifier(testID)
    }

    // MARK: - Sheet Content

    private var sheetContent: some View {
        VStack(alignment: .leading, spacing: theme.space.md) {
            // Info icon
            HStack {
                Spacer()
                LSIconSymbol(
                    name: "information",
                    size: 32,
                    color: theme.colors.primary.default
                )
                .padding(theme.space.md)
                .background(
                    Circle()
                        .fill(theme.colors.primary.default.opacity(0.15))
                )
                Spacer()
            }

            // Title
            Text("Favorites Not Included")
                .font(theme.type.title.md.font)
                .foregroundStyle(theme.colors.onSurface.default)
                .frame(maxWidth: .infinity)
                .multilineTextAlignment(.center)
                .accessibilityIdentifier("\(testID)-title")

            // Message
            Text("These favorite roads are too far from your planned route:")
                .font(theme.type.body.md.font)
                .foregroundStyle(theme.colors.onSurface.muted ?? theme.colors.onSurface.default.opacity(0.6))
                .accessibilityIdentifier("\(testID)-message")

            // Favorites list
            if !unavailableFavorites.isEmpty {
                VStack(alignment: .leading, spacing: theme.space.sm) {
                    ForEach(unavailableFavorites, id: \.self) { favorite in
                        Text("• \(favorite)")
                            .font(theme.type.body.md.font)
                            .foregroundStyle(theme.colors.onSurface.default)
                            .accessibilityLabel(favorite)
                    }
                }
                .accessibilityElement(children: .contain)
                .accessibilityLabel("Favorite roads not included")
                .padding(theme.space.md)
                .background(
                    RoundedRectangle(cornerRadius: theme.radius.md)
                        .fill(theme.colors.surface.default.opacity(0.5))
                )
                .accessibilityIdentifier("\(testID)-list")
            }

            // Guidance
            Text("Try planning a route nearer to these favorites, or add them to a different route.")
                .font(theme.type.body.sm.font)
                .foregroundStyle(theme.colors.onSurface.muted ?? theme.colors.onSurface.default.opacity(0.6))
                .accessibilityIdentifier("\(testID)-guidance")

            // Got it button
            LSButton(
                "Got it",
                variant: .default,
                size: .lg,
                onPress: onDismiss,
                testID: "\(testID)-close-button"
            )
        }
        .padding(theme.space.lg)
        .presentationDetents([.presentation(0.6)])
        .presentationDragIndicator(.visible)
    }
}

// MARK: - Preview

#Preview("FavoritesInfoSheet - Shown") {
    struct PreviewWrapper: View {
        @State private var isVisible = true

        var body: some View {
            VStack {
                Text("Parent View")
                    .font(.headline)

                LSFavoritesInfoSheet(
                    visible: isVisible,
                    unavailableFavorites: [
                        "Pacific Coast Highway",
                        "Mulholland Drive",
                        "Sunset Boulevard",
                    ],
                    onDismiss: {
                        isVisible = false
                    },
                    testID: "preview-favorites-info-sheet"
                )
            }
        }
    }

    return PreviewWrapper()
        .laneShadowTheme()
}

#Preview("FavoritesInfoSheet - Empty List") {
    struct PreviewWrapper: View {
        @State private var isVisible = true

        var body: some View {
            VStack {
                Text("Parent View")
                    .font(.headline)

                LSFavoritesInfoSheet(
                    visible: isVisible,
                    unavailableFavorites: [],
                    onDismiss: {
                        isVisible = false
                    },
                    testID: "preview-empty-sheet"
                )
            }
        }
    }

    return PreviewWrapper()
        .laneShadowTheme()
}

#Preview("FavoritesInfoSheet - Light Mode") {
    struct PreviewWrapper: View {
        @State private var isVisible = true

        var body: some View {
            LSFavoritesInfoSheet(
                visible: isVisible,
                unavailableFavorites: ["Test Road"],
                onDismiss: {
                    isVisible = false
                }
            )
        }
    }

    return PreviewWrapper()
        .laneShadowTheme()
        .preferredColorScheme(.light)
}

#Preview("FavoritesInfoSheet - Dark Mode") {
    struct PreviewWrapper: View {
        @State private var isVisible = true

        var body: some View {
            LSFavoritesInfoSheet(
                visible: isVisible,
                unavailableFavorites: ["Test Road"],
                onDismiss: {
                    isVisible = false
                }
            )
        }
    }

    return PreviewWrapper()
        .laneShadowTheme()
        .preferredColorScheme(.dark)
}
